website/integrations: Improve discord role sync policies

Keyinator opened the pull request:
## Details

This changes the websites integration page for discord by modifying two policies.

The new policies are easier to understand and take advantage of group attributes instead of hardcoding values (besides the guild id).

---

## Checklist

-   [ ] Local tests pass (`ak test authentik/`)
-   [ ] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ ] The code has been formatted (`make web`)

If applicable

-   [X] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

Keyinator committed:
Improve discord role sync policies

Keyinator committed:
Improvements to discord role sync policies

- Add descriptions
- Fix spelling
- Improved Code styling

burritosoftware commented:
I tested these last week with @Keyinator over Discord and they work great!

tanberry commented on website/docs/sources/discord/index.md:
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
 :::
 
+:::info
+Any authentik role you want to sync with a discord role needs to have the attribute `discord_role_id` with a value of the discord role's id set.  
```
```suggestion
Any authentik role that you want to sync with a Discord role needs to have the attribute `discord_role_id` with a value of the Discord role's ID set.  
```

tanberry commented on website/docs/sources/discord/index.md:
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
+Blueprint attribute: ``discord_role_id: "<ROLE ID>"``
+:::
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.  
```
```suggestion
The following two policies allow you to synchronize roles in a Discord guild with roles in authentik.  
```

tanberry commented on website/docs/sources/discord/index.md:
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
+:::
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.  
+Whenever a user enrolls or signs in to authentik via discord source, these policies will check the user's discord roles and apply the user's authentik roles accordingly.  
```
```suggestion
Whenever a user enrolls or signs in to authentik via a Discord source, these policies will check the user's Discord roles and apply the user's authentik roles accordingly.  
```

tanberry commented on website/docs/sources/discord/index.md:
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
+
+The following two policies will allow you to synchronize roles in a discord guild with roles in authentik.  
+Whenever a user enrolls or signs in to authentik via discord source, these policies will check the user's discord roles and apply the user's authentik roles accordingly.  
+All roles with the attribute ``discord_role_id`` defined will be added or removed depending on whether the user is a member of the defined discord role.
```
```suggestion
All roles with the attribute ``discord_role_id`` defined will be added or removed depending on whether the user is a member of the defined Discord role.
```

tanberry reviewed (changes requested):
Thanks for this contribution @Keyinator ! A couple of nits about capitalizing Discord, but yeah as soon as those are made let's merge!

Keyinator committed:
Apply suggestions from code review

Co-authored-by: Tana M Berry <tanamarieberry@yahoo.com>
Signed-off-by: Keyinator <k3yinator@gmail.com>

Keyinator committed:
Merge branch 'goauthentik:main' into main

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -158,35 +158,36 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
 # Ensure flow is only run during OAuth logins via Discord
-if context["source"].provider_type != "discord":
-    return True
+if context['source'].provider_type != "discord":
```
```suggestion
if context["source"].provider_type != "discord":
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -195,28 +196,36 @@ if not connection:
 # Ensure user is a member of the guild
 if "code" in guild_member_info:
-    if guild_member_info["code"] == 10004:
+    if guild_member_info['code'] == 10004:
```
```suggestion
    if guild_member_info["code"] == 10004:
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -195,28 +196,36 @@ if not connection:
-    headers={"Authorization": "Bearer " + access_token},
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": "Bearer " + access_token
```
```suggestion
    "Authorization": f"Bearer {access_token}",
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -255,35 +255,44 @@ if not connection:
-        ak_create_event(
-            "discord_error", source=context["source"], code=guild_member_info["code"]
-        )
+        ak_create_event("discord_error", source=context['source'], code=guild_member_info['code'])
```
```suggestion
        ak_create_event("discord_error", source=context["source"], code=guild_member_info["code"])
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -195,28 +196,36 @@ if not connection:
-        ak_create_event(
-            "discord_error", source=context["source"], code=guild_member_info["code"]
-        )
+        ak_create_event("discord_error", source=context['source'], code=guild_member_info['code'])
```
```suggestion
        ak_create_event("discord_error", source=context["source"], code=guild_member_info["code"])
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -255,35 +255,44 @@ if not connection:
 # Update user's groups
-request.user.ak_groups.set(new_groups)
+request.user.ak_groups.set(combined_groups)
 request.user.save()
```
Saving the user here should not be needed

rissson replied:
```suggestion
```

Keyinator replied:
You are right.

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -195,28 +196,36 @@ if not connection:
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
```
```suggestion
matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info["roles"])
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -195,28 +196,36 @@ if not connection:
+# Create event with roles added
+ak_create_event(
+    "discord_role_sync",
+    discord_roles_added=', '.join(str(group) for group in matching_roles),
```
```suggestion
    discord_roles_added=", ".join(str(group) for group in matching_roles),
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -255,35 +255,44 @@ if not connection:
+discord_groups = Group.objects.filter(attributes__discord_role_id__isnull=False)
+
+# Get all user groups except discord_groups
+user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list('pk', flat=True))
```
```suggestion
user_groups = request.user.ak_groups.exclude(pk__in=discord_groups.values_list("pk", flat=True))
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -255,35 +255,44 @@ if not connection:
-    if role_id in guild_member_info["roles"]:
-        new_groups.append(MAPPED_ROLES[role_id])
+# Filter matching roles based on guild_member_info['roles']
+matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info['roles'])
```
```suggestion
matching_roles = discord_groups.filter(attributes__discord_role_id__in=guild_member_info["roles"])
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -255,35 +255,44 @@ if not connection:
+# Create event with roles changed
+ak_create_event(
+    "discord_role_sync",
+    discord_roles_before=', '.join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))),
+    discord_roles_after=', '.join(str(group) for group in matching_roles),
```
```suggestion
    discord_roles_before=", ".join(str(group) for group in request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))),
    discord_roles_after=", ".join(str(group) for group in matching_roles),
```

rissson replied:
Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

Keyinator replied:
> Those queries should already be available from above. Why not reuse them instead of doing additional database queries?

Do you mean ``request.user_ak_groups`` or ``request.user.ak_groups.filter(pk__in=discord_groups.values_list('pk', flat=True))``?

If you are saying that ``request.user.ak_groups`` is calculated on every access, we can definitely do that.
If you are referring to the value of ``discord_roles_before`` then the following should explain why we can't reuse above code:

In the code preceeding the event we filter out roles which are __not__ in `discord_groups`.
However in this event, we want to display `discord_groups` the user was in.

Basically we have these relations:
`User-Roles` - `discord_groups` (preceeding code)
`User-Roles` ∩ `discord_groups` (within event)

If I am not mistaken, one cannot easily be reused for the other

Please let me know which you mean so I can adjust the code.

rissson replied:
So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

Keyinator replied:
> So actually, the second one is fine, but the first one, `discord_roles_before` is I think wrong. As written, the query is based on `request.user.ak_groups` _after_ it's already been modified. So if any roles were removed, they wouldn't show up in the event.

I checked and it seems to be working fine.
I suppose the request-object is just the in-time data when the policy is executed.

rissson replied:
No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

Keyinator replied:
> No, it will actually execute the `request.user.ak_groups.filter(pk__in=discord_groups.values_list("pk", flat=True))` query after you've modified the data. If we really want to log proper data, that query should be done before changing ak_groups, saved as a variable and used when needed.

I tested it on 2024.4.3 and ``discord_roles_before`` was showing correctly. But we can change it if you prefer.

rissson replied:
Yeah, I'd rather we change it, so we're 100% sure it works

rissson committed:
Apply suggestions from code review

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

Keyinator committed:
Apply code style guide

Keyinator committed:
Handle discord ratelimiting

Keyinator committed:
Remove unnecessary request.user.save()

Co-authored-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
Signed-off-by: Keyinator <k3yinator@gmail.com>

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -202,6 +202,11 @@ guild_member_info = requests.get(
 ).json()
 
+# Ensure we are not being ratelimited
+if guild_member_info.get("message") == "You are being rate limited.":
```
This is quite weak as a check. It would be better to:

```python3
guild_member_request = requests.get(
  GUILD_API_URL.format(guild_id=guild_id),
  headers={
    "Authorization": f"Bearer {access_token}",
  },
)
guild_member_info = guild_member_request.json()

# Ensure we are not being ratelimited
if guild_member_request.status_code == 429:
	pass # do things
```

And the same below.

Cf. https://discord.com/developers/docs/topics/rate-limits

Keyinator replied:
I am not sure if guild_member_request has this info. When I sent guild_member_info as an event it only had the following properties:
```
"guild_member_info": {
        "global": false,
        "message": "You are being rate limited.",
        "retry_after": 35.683
    },
```

rissson replied:
So it goes like this:
```python
guild_member_request = requests.get(...)
# This is a requests.Response object, which has, among a lot of other things, a status_code attribute and a json() method.
guild_member_info = guild_member_request.json()
# This takes the requests.Response.body and passes it to json.loads and gives it back.
```
So from `guild_member_request`, you can get the HTTP status code to check you're not being rate limited, and then from `guild_member_info` you can get either the "roles" key (and others) if you're not being rate limited, or the `retry_after` key if you are.

Keyinator committed:
Improvements

- Improve resilience by using http status instead of message and evaluating variables earlier
- Improve readability by using more descriptive variable names

tanberry commented:
HI @Keyinator can you run `make website` or `npm run prettier` again and repush? That build check is failing. We'll get this merged yet!! :-)

Keyinator committed:
Run prettier

Keyinator commented:
> HI @Keyinator can you run `make website` or `npm run prettier` again and repush? That build check is failing. We'll get this merged yet!! :-)

There were indeed fails in my file (fixed now).  
However there is also lint-issues in `website/integrations/services/gitlab/index.md`. How do we proceed?

rissson commented:
> There were indeed fails in my file (fixed now). However, there is also lint-issues in `website/integrations/services/gitlab/index.md`. How do we proceed?

You can `git fetch && git merge origin/main`, which should fix the issue in this particular branch. I suppose it was created when we had those fails on `main`.

Keyinator committed:
Merge branch 'goauthentik:main' into main

rissson commented:
I'm sorry we broke the main CI again (we're having GitHub actions issues with this currently). #10524 should fix it. Once that's merged, could you merge main into this branch again?

Keyinator commented:
> I'm sorry we broke the main CI again (we're having GitHub actions issues with this currently). #10524 should fix it. Once that's merged, could you merge main into this branch again?

Sure thing :)

Keyinator committed:
Merge branch 'goauthentik:main' into main

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
 # Ensure flow is only run during OAuth logins via Discord
 if context["source"].provider_type != "discord":
-    return True
+  return True
```
```suggestion
    return True
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
-    headers={"Authorization": "Bearer " + access_token},
-).json()
+guild_member_request = requests.get(
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": f"Bearer {access_token}",
+  },
```
```suggestion
    GUILD_API_URL.format(guild_id=guild_id),
    headers={
        "Authorization": f"Bearer {access_token}",
    },
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -158,65 +158,80 @@ Now bind this policy to the chosen enrollment and authentication flows for the D
+
+# Ensure we are not being ratelimited
+if guild_member_request.status_code == 429:
+  ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
+  return False
```
```suggestion
    ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
    return False
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -254,35 +260,50 @@ if not connection:
-    headers={"Authorization": "Bearer " + access_token},
-).json()
+guild_member_request = requests.get(
+  GUILD_API_URL.format(guild_id=guild_id),
+  headers={
+    "Authorization": f"Bearer {access_token}"
+  },
+)
```
```suggestion
    GUILD_API_URL.format(guild_id=guild_id),
    headers={
        "Authorization": f"Bearer {access_token}"
    },
)
```

rissson commented on website/docs/sources/discord/index.md:
```diff
@@ -254,35 +260,50 @@ if not connection:
+
+# Ensure we are not being ratelimited
+if guild_member_request.status_code == 429:
+  ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
+  return False
```
```suggestion
    ak_message(f"Discord is throttling this connection. Retry in {int(guild_member_info['retry_after'])}s")
    return False
```

rissson committed:
Update website/docs/sources/discord/index.md

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Update website/docs/sources/discord/index.md

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Update website/docs/sources/discord/index.md

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Update website/docs/sources/discord/index.md

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>

rissson committed:
Update website/docs/sources/discord/index.md

Signed-off-by: Marc 'risson' Schmitt <marc.schmitt@risson.space>
